### PR TITLE
Tokenizers v3.0.0

### PR DESCRIPTION
**Pull request by mfuntowicz**

### CODE REVIEW

File: tests/test_tokenization_fast.py [hunk 1]
```diff
@@ -21,117 +24,92 @@
...
-        self.assertLessEqual(inputs_diff / a_.shape[0], threshold)
+    def setUp(self) -> None:
+        with open("tests/fixtures/sample_text.txt", encoding="utf-8") as f_data:
+            self._data = f_data.read().replace("\n\n", "\n").strip()
 
-    def assert_tokenization_python_rust_almost_equals(self, tokenizer_p, tokenizer_r, threshold: float):
+    def test_fast_align_python(self):
+        for (name, rust_cls, python_cls, vocab_key) in self.TOKENIZERS_CLASSES:
+            for pretrained_name in python_cls.pretrained_vocab_files_map[vocab_key].keys():
+                with self.subTest("{} ({})".format(name, pretrained_name)):
```
Comment (LysandreJik) [inline tests/test_tokenization_fast.py]: This seems to me like to optimal organisation. Testing on every checkpoint on every tokenizer organized in subTests is really thorough.

Does it take a while? Should we mark this as slow or is it fast enough?

File: tests/test_tokenization_fast.py [hunk 2]
```diff
@@ -258,281 +238,147 @@ def assert_build_inputs_with_special_tokens(self, tokenizer_r, tokenizer_p):
...
-            # Check alignment for build_inputs_with_special_tokens
-            self.assert_build_inputs_with_special_tokens(tokenizer_r, tokenizer_p)
+            # Ensure we match max_length
+            self.assertEqual(len(input_r), max_length), self.assertEqual(len(input_p), max_length)
 
-            # Check the number of returned files for save_vocabulary
-            self.assert_save_pretrained(tokenizer_r, tokenizer_p)
+            # Ensure the number of padded tokens is the same
+            padded_tokens_r = list(takewhile(lambda i: i == tokenizer_r.pad_token_id, reversed(input_r)))
+            padded_tokens_p = list(takewhile(lambda i: i == tokenizer_p.pad_token_id, reversed(input_p)))
```
Comment (LysandreJik) [inline tests/test_tokenization_fast.py]: We usually try to use lambdas as little as possible, as they're usually a bit hard to read. cc @thomwolf

**Review by LysandreJik — APPROVED** [review]
Overall, this is quite a different test suite than what we have in `test_modeling_common` and `test_tokenization_common`, but in a good way, imo.

The use of subTest will greatly help debugging and splitting between `test_xxx` methods and `assert_xxx` methods makes the code cleaner and easier to read.

File: tests/test_tokenization_fast.py (hunk 2 above)
Comment (julien-c) [reply tests/test_tokenization_fast.py]: hmm here I think it's fine, no?

Comment (thomwolf) [reply tests/test_tokenization_fast.py]: In tests I'm more fine with them

Comment (mfuntowicz) [reply tests/test_tokenization_fast.py]: 👍

File: src/transformers/tokenization_utils.py [hunk 3]
```diff
@@ -1950,51 +1953,67 @@ def batch_encode_plus(
...
-            return BatchEncoding(
-                {
-                    key: value[0] if len(value) > 0 and isinstance(value[0], list) else value
-                    for key, value in batched_output.items()
-                },
-                batched_output.encodings,
-            )
+        # Check for pretokenized path (ie [token1, token2, ..., tokenN] -> [id1, id2, ..., idN]
+        if isinstance(text, list) and text_pair is None:
+            encoding = self._tokenizer.encode_tokenized(text)
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py]: I think we should have a `encode_tokenized_batch` in Rust and add a `is_pretrokenized` flag to `encode_plus` and `batch_encode_plus` as discussed

File: src/transformers/tokenization_utils.py [hunk 4]
```diff
@@ -1820,48 +1821,50 @@ def _convert_token_to_id_with_added_voc(self, token):
...
-        stride=0,
-        truncation_strategy="longest_first",
-        pad_to_max_length=False,
-        return_tensors=None,
-        return_token_type_ids=True,
-        return_attention_mask=True,
-        return_overflowing_tokens=False,
-        return_special_tokens_mask=False,
-        return_offsets_mapping=False,
+        batch_text_or_text_pairs: Union[List[str], List[Tuple[str, str]]] = None,
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py]: Let's allow batch pretokenized inputs and change this to `Union[List[str], List[List[str]],, List[Tuple[str, str]], List[Tuple[List[str], List[str]]]]`?

Comment (thomwolf) [reply src/transformers/tokenization_utils.py]: I think we can assume that if the inputs are sentence pairs, both members of the pairs will be either not pretokenized or both pretokenized.

File: src/transformers/tokenization_utils.py [hunk 5]
```diff
@@ -1950,51 +1953,67 @@ def batch_encode_plus(
...
+        text: Union[str, List[str]],
+        text_pair: Optional[str] = None,
+        add_special_tokens: bool = True,
+        max_length: Optional[int] = None,
+        pad_to_max_length: bool = False,
+        stride: int = 0,
+        truncation_strategy: str = "longest_first",
+        return_tensors: Optional[str] = None,
+        return_token_type_ids: bool = True,
+        return_attention_mask: bool = True,
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py]: You should rebase on master if you can because I think this is now Optional[bool] = None since @LysandreJik worked to adapt the output to the models.

**Review by thomwolf — COMMENTED** [review]
This is looking great, just a few comments on things to check

File: src/transformers/tokenization_utils.py (hunk 5 above)
Comment (mfuntowicz) [reply src/transformers/tokenization_utils.py]: I'm not sure to fully undestand why bool should be Optional ? A default value would be more understable imho and remove the need for None checking, wdyt ?

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (mfuntowicz) [reply src/transformers/tokenization_utils.py]: How do we want to handle pairs here ? On Rust side, we have to provide the a list of first tuples part and then a second one: 

```python
# First part
enc_first = encode_tokenized_batch([s for (s, p) in batch_text_or_text_pairs], 0)

# Second part
enc_second = encode_tokenized_batch([p for (s, p) in batch_text_or_text_pairs], 1)
```

Then we have two BatchEncoding, unrelated to each other (@n1t0 correct me if I'm wrong here), do we want to unify them ? Return both ?

File: src/transformers/tokenization_utils.py (hunk 5 above)
Comment (julien-c) [reply src/transformers/tokenization_utils.py]: You should check w @LysandreJik :)

Comment (LysandreJik) [reply src/transformers/tokenization_utils.py]: The spirit behind having those values as `None` by default is the following: if this value is `None`, then all those returns are set to the default *tokenizer-specific values*. This is different from tokenizer to tokenizer, e.g. DistilBERT should by default return `input_ids` and `attention_mask`, but **not** `token_type_ids`, as the model cannot handle it. This in turn allows the user to do the following:

```py
inputs = tokenizer.encode_plus(values, return_tensors="pt")
model(**inputs)
```

And this now works with every model.

These values can still be explicitly set to `True` or `False` by the user. See #3116 for more information/implementation details.

File: tests/test_tokenization_fast.py (hunk 1 above)
Comment (mfuntowicz) [reply tests/test_tokenization_fast.py]: It takes around 1 min for doing everything

**Review by thomwolf — COMMENTED** [review]
Ok, I think there are still a lot of things to clean up but I'm starting to find this PR really hard to read. Probably time to try to merge with a rather clean state and start a PR to clean all this up.

File: src/transformers/tokenization_utils.py [hunk 6]
```diff
@@ -103,44 +105,62 @@ def truncate_and_pad(
...
+            raise ValueError("word_boundaries() is not available when using Python based tokenizers")
+        return self[sample].word_boundaries(pos)
 
-        - ``additional_special_tokens``: (`Optional`) list: a list of additional special tokens. Adding all special tokens here ensure they won't be split by the tokenization process. Will be associated to ``self.additional_special_tokens`` and ``self.additional_special_tokens_ids``
-    """
+    def token_to_word(self, sample: int, index: int) -> Optional[Tuple[int, int]]:
+        """ Find the offsets of the word that contains the token at the given index """
+        if not self._encodings:
+            raise ValueError("token_to_word() is not available when using Python based tokenizers")
+        return self[sample].word_boundaries(index)
```
[the remark this answers is not in the document]
Comment (mfuntowicz) [reply src/transformers/tokenization_utils.py]: It's relics of bad copy/paste

File: src/transformers/tokenization_transfo_xl.py [hunk 7]
```diff
@@ -404,13 +404,26 @@ def __init__(
...
+            [seq.strip() if isinstance(seq, str) else (seq[0].strip(), seq[1].strip()) for seq in sequences],
+            add_special_tokens=add_special_tokens,
         )
 
-    def encode(self, sequence: str, pair: Optional[str] = None) -> Encoding:
-        return super().encode(sequence.strip(), pair.strip() if pair else pair)
+    def encode(self, sequence: str, pair: Optional[str] = None, add_special_tokens: bool = False) -> Encoding:
+        if sequence is None:
+            raise ValueError(
+                "Input is not valid. Should be a string, a list/tuple of strings or a list/tuple of integers."
```
Comment (n1t0) [inline src/transformers/tokenization_transfo_xl.py]: This one should probably just say `string`

File: src/transformers/tokenization_utils.py [hunk 8]
```diff
@@ -1785,7 +1911,7 @@ def _convert_encoding(
             if return_special_tokens_mask:
                 encoding_dict["special_tokens_mask"].append(e.special_tokens_mask)
             if return_offsets_mapping:
-                encoding_dict["offset_mapping"].append([e.original_str.offsets(o) for o in e.offsets])
+                encoding_dict["offset_mapping"] = [o for o in e.offsets]
```
Comment (n1t0) [inline src/transformers/tokenization_utils.py]: Should it be like above:
```suggestion
                encoding_dict["offset_mapping"].append(e.offsets)
```

File: src/transformers/tokenization_utils.py [hunk 9]
```diff
@@ -1818,55 +1944,57 @@ def _convert_token_to_id_with_added_voc(self, token):
...
-        if not add_special_tokens:
-            logger.warning(
-                "Fast tokenizers add special tokens by default. To remove special tokens, please specify"
-                "`add_special_tokens=False` during the initialisation rather than when calling `encode`,"
-                "`encode_plus` or `batch_encode_plus`."
+    ) -> BatchEncoding:
+
+        if batch_text_or_text_pairs is None:
+            raise ValueError(
+                "Input is not valid. Should be a string, a list/tuple of strings or a list/tuple of integers."
```
Comment (n1t0) [inline src/transformers/tokenization_utils.py]: Same here, I think we can remove the `string` option

File: tests/test_tokenization_fast.py [hunk 10]
```diff
@@ -21,118 +21,114 @@
...
+
+        # Check that Rust and Python align
+        self.assert_tokenization_python_rust_equals(tokenizer_r, tokenizer_p)
+        self.assert_num_special_tokens_to_add_equal(tokenizer_r, tokenizer_p)
+        self.assert_max_length_equal(tokenizer_r, tokenizer_p)
+        self.assert_special_tokens_map_equal(tokenizer_r, tokenizer_p)
+        self.assert_embeded_special_tokens(tokenizer_r, tokenizer_p)
+        self.assert_padding(tokenizer_r, tokenizer_p)
+        # TODO: enable for v3.0.0
+        # self.assert_empty_output_no_special_tokens(tokenizer_r, tokenizer_p)
```
Comment (n1t0) [inline tests/test_tokenization_fast.py]: Should this be enabled?

File: tests/test_tokenization_fast.py [hunk 11]
```diff
@@ -258,281 +254,275 @@ def assert_build_inputs_with_special_tokens(self, tokenizer_r, tokenizer_p):
...
-            # Check for offsets mapping
-            self.assert_offsets_mapping(tokenizer_r)
+    def assert_embeded_special_tokens(self, tokenizer_r, tokenizer_p):
+        sentence = "A, <mask> AllenNLP sentence."
+        tokens_r = tokenizer_r.encode_plus(sentence, add_special_tokens=True, return_token_type_ids=True)
+        tokens_p = tokenizer_p.encode_plus(sentence, add_special_tokens=True, return_token_type_ids=True)
 
-            # Check for dynamic encoding sequence handling in batch_encode_plus
-            self.assert_batch_encode_dynamic_overflowing(tokenizer_r)
+        # Rust correctly handles the space before the mask while python doesnt
```
Comment (n1t0) [inline tests/test_tokenization_fast.py]: I think Python is right here. There shouldn't be any space before the `<mask>` token. This means that Roberta on the fast path should probably have an `AddedToken('<mask>', lstrip=True)`

File: tests/test_tokenization_gpt2.py [hunk 12]
```diff
@@ -94,18 +94,18 @@ def test_rust_and_python_full_tokenizers(self):
...
         tokenizer = self.get_tokenizer()
-        rust_tokenizer = self.get_rust_tokenizer(add_special_tokens=False, add_prefix_space=True)
+        rust_tokenizer = self.get_rust_tokenizer(add_prefix_space=True)
 
         sequence = "lower newer"
 
         # Testing tokenization
         tokens = tokenizer.tokenize(sequence, add_prefix_space=True)
-        rust_tokens = rust_tokenizer.tokenize(sequence)
+        rust_tokens = rust_tokenizer.tokenize(sequence, add_prefix_space=True)
```
Comment (n1t0) [inline tests/test_tokenization_gpt2.py]: `add_prefix_space=True` isn't required here I think

**Review by n1t0 — COMMENTED** [review]
A few details here and there, but otherwise looks good to me!

File: src/transformers/tokenization_utils.py [hunk 13]
```diff
@@ -103,44 +123,100 @@ def truncate_and_pad(
...
+    def items(self):
+        return self.data.items()
 
-        - ``mask_token``: (`Optional`) string: a masking token (e.g. when training a model with masked-language modeling). Will be associated to ``self.mask_token`` and ``self.mask_token_id``
+    def char_to_token_offsets(self, sentence: int, char: int) -> Tuple[int, int]:
+        """
+        Find the Offsets of the token containing the character at the specified position
+        :param sentence: Index of the sentence relative to the batch provided to the tokenizer.
+        :param char: Char index to get the relative token offsets
+        :return: (token start, token end)
```
Comment (LysandreJik) [inline src/transformers/tokenization_utils.py:175]: (Applicable to most other docstrings) We use [Google style doc](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) in the library - could we try and use them here as well?

**Review by LysandreJik — APPROVED** [review]
This isn't easy to review as the diff mixes substractions from `PreTrainedTokenizer` and additions from `BatchEncoding`; other than that, cool! Thanks @mfuntowicz :)

**Comment by LysandreJik** [comment]
Really like the new typings!

File: src/transformers/pipelines.py [hunk 14]
```diff
@@ -459,7 +459,7 @@ def _parse_and_tokenize(self, *texts, pad_to_max_length=False, **kwargs):
         )
 
         # Filter out features not available on specific models
-        inputs = self.inputs_for_model(inputs)
+        # inputs = self.inputs_for_model(inputs)
```
Comment (julien-c) [inline src/transformers/pipelines.py:462]: Let's remember to remove this for good soon @mfuntowicz
